File: packages/zed-editor-bin/update.py
#!/usr/bin/env python3
"""Update Zed (aarch64-darwin) from the official stable release API."""

from __future__ import annotations

import argparse
import json
import re
import subprocess
import sys
import urllib.error
import urllib.request
from pathlib import Path

ROOT = Path(__file__).parents[2]
RELEASE_URL = (
    "https://cloud.zed.dev/releases/stable/latest/asset"
    "?asset=zed&os=macos&arch=aarch64"
)
USER_AGENT = "9bingyin-nur-packages-updater"


def run(command: list[str], *, capture: bool = False) -> str:
    result = subprocess.run(command, check=True, text=True, capture_output=capture)
    return result.stdout if capture else ""


def latest_release() -> tuple[str, str]:
    request = urllib.request.Request(
        RELEASE_URL,
        headers={"User-Agent": USER_AGENT},
    )
    with urllib.request.urlopen(request, timeout=30) as response:
        release: object = json.load(response)

    if not isinstance(release, dict):
        raise RuntimeError("Zed release API returned an invalid payload")

    version = release.get("version")
    url = release.get("url")
    if not isinstance(version, str) or not re.fullmatch(
        r"[0-9]+(?:\.[0-9]+)+", version
    ):
        raise RuntimeError("Zed release API returned an invalid version")

    expected_url = (
        "https://github.com/zed-industries/zed/releases/download/"
        f"v{version}/Zed-aarch64.dmg"
    )
    if url != expected_url:
        raise RuntimeError(f"Zed release API returned an unexpected URL: {url!r}")

    return version, expected_url


def prefetch_sri_hash(url: str) -> str:
    payload: object = json.loads(
        run(["nix", "store", "prefetch-file", "--json", url], capture=True)
    )
    if not isinstance(payload, dict):
        raise RuntimeError(f"nix store prefetch-file returned an invalid payload for {url}")

    hash_value = payload.get("hash")
    if not isinstance(hash_value, str):
        raise RuntimeError(f"nix store prefetch-file returned no hash for {url}")
    return hash_value


def replace_once(text: str, pattern: str, replacement: str, error: str) -> str:
    updated, count = re.subn(pattern, replacement, text, count=1, flags=re.MULTILINE)
    if count != 1:
        raise RuntimeError(error)
    return updated


def current_version(package_text: str) -> str:
    match = re.search(r'^  version = "([^"]+)";', package_text, flags=re.MULTILINE)
    if match is None:
        raise RuntimeError("Failed to read the current Zed version")
    return match.group(1)


def update_package(version: str, url: str) -> None:
    package_path = ROOT / "packages/zed-editor-bin/package.nix"
    text = package_path.read_text()
    if current_version(text) == version:
        print(f"zed-editor-bin is already at {version}")
        return

    text = replace_once(
        text,
        r'^  version = "[^"]+";',
        f'  version = "{version}";',
        "Failed to update the Zed version",
    )
    text = replace_once(
        text,
        r'(hash = ")[^"]+(";)',
        rf"\g<1>{prefetch_sri_hash(url)}\2",
        "Failed to update the Zed hash",
    )
    package_path.write_text(text)


def main() -> None:
    argparse.ArgumentParser(description=__doc__).parse_args()
    version, url = latest_release()
    update_package(version, url)


if __name__ == "__main__":
    try:
        main()
    except (
        OSError,
        RuntimeError,
        subprocess.CalledProcessError,
        urllib.error.URLError,
    ) as error:
        print(f"error: {error}", file=sys.stderr)
        raise SystemExit(1) from error

File: packages/grok-bot/update.ts
#!/usr/bin/env bun
/** Update Grok Bot (aarch64-darwin) from Cursor's stable update feed. */

import { join } from "node:path";
import process from "node:process";

const ARCH = "darwin-arm64";
const FEED_URL = `https://api2.cursor.sh/updates/api/update/${ARCH}/sand/0.0.1/stable`;
const DOWNLOAD_BASE = "https://downloads.cursor.com/grokbot/stable";
const PACKAGE_PATH = join(import.meta.dir, "package.nix");
const USER_AGENT = "9bingyin-nur-packages-updater";
const VERSION_RE = /^[0-9]+(?:\.[0-9]+)+$/;

function isRecord(value: unknown): value is Record<string, unknown> {
	return typeof value === "object" && value !== null && !Array.isArray(value);
}

function run(command: readonly string[], capture = false): string {
	if (command.length === 0) {
		throw new Error("Command must not be empty");
	}
	const result = Bun.spawnSync([...command], {
		stderr: "inherit",
		stdout: capture ? "pipe" : "inherit",
	});
	if (result.exitCode !== 0) {
		throw new Error(
			`Command failed with exit code ${result.exitCode}: ${command.join(" ")}`,
		);
	}
	return capture ? result.stdout.toString() : "";
}

async function latestRelease(): Promise<readonly [string, string]> {
	const response = await fetch(FEED_URL, {
		headers: { "User-Agent": USER_AGENT },
		signal: AbortSignal.timeout(30_000),
	});
	if (!response.ok) {
		throw new Error(
			`Grok Bot update feed returned HTTP ${response.status} ${response.statusText}`,
		);
	}
	const payload: unknown = await response.json();
	if (!isRecord(payload)) {
		throw new Error("Grok Bot update feed returned a non-object payload");
	}
	const version = payload.name;
	const url = payload.url;
	if (typeof version !== "string" || !VERSION_RE.test(version)) {
		throw new Error("Grok Bot update feed has an invalid version");
	}
	const expectedUrl = `${DOWNLOAD_BASE}/${ARCH}/${version}/Grok_Bot_${version}.zip`;
	if (url !== expectedUrl) {
		throw new Error(
			`Grok Bot update feed has an unexpected download URL: ${JSON.stringify(url)}`,
		);
	}
	return [version, expectedUrl];
}

function prefetchSriHash(url: string): string {
	const payload: unknown = JSON.parse(
		run(["nix", "store", "prefetch-file", "--json", url], true),
	);
	if (!isRecord(payload)) {
		throw new Error(
			`nix store prefetch-file returned an invalid payload for ${url}`,
		);
	}
	const hashValue = payload.hash;
	if (typeof hashValue !== "string") {
		throw new Error(`nix store prefetch-file returned no hash for ${url}`);
	}
	return hashValue;
}

function replaceOnce(
	text: string,
	pattern: string,
	replacement: string,
	error: string,
): string {
	const updated = text.replace(new RegExp(pattern, "m"), replacement);
	if (updated === text) {
		throw new Error(error);
	}
	return updated;
}

function currentVersion(packageText: string): string {
	const match = /^ {2}version = "([^"]+)";/m.exec(packageText);
	if (match?.[1] === undefined) {
		throw new Error("Failed to read current Grok Bot version");
	}
	return match[1];
}

async function updatePackage(version: string, url: string): Promise<void> {
	const packageText = await Bun.file(PACKAGE_PATH).text();
	if (currentVersion(packageText) === version) {
		console.log(`grok-bot is already at ${version}`);
		return;
	}

	const updated = replaceOnce(
		replaceOnce(
			packageText,
			'^  version = "[^"]+";',
			`  version = "${version}";`,
			"Failed to update Grok Bot version",
		),
		'(hash = ")[^"]+(";)',
		`$1${prefetchSriHash(url)}$2`,
		"Failed to update Grok Bot hash",
	);
	await Bun.write(PACKAGE_PATH, updated);
}

async function main(): Promise<void> {
	const [version, url] = await latestRelease();
	await updatePackage(version, url);
}

try {
	await main();
} catch (error) {
	const message = error instanceof Error ? error.message : String(error);
	console.error(`error: ${message}`);
	process.exitCode = 1;
}
